### PR TITLE
move uniform_ from TH to ATen and optimize with MKL VSL

### PR DESCRIPTION
**Pull request by mingfeima**
This PR moves `uniform_` CPU implementation from TH to ATen, e.g. `uniform_cpu_`.
When PyTorch is built with MKL support and CPU is deteced to be intel CPU, `uniform_cpu_` will finally call MKL VSL random generator which assures best performance on intel CPUs.
Otherwise (MKL is not built or AMD CPU), `uniform_cpu_` will end up in sequential random generator from ATen, this path is identical to old TH implementation.

Single socket run speedup: **2.4x~142x**
Single core run speedup: **3.2x~13x**

### CODE REVIEW

**Comment by mingfeima** [comment]
### performance
Below shows the performance compare between original and this pr, the benchmark will compare from:

1. a range of input size: [1000], [10000], [100000], [1000000]
2. single socket and single core

benchmark code available at this [op_bench-py](https://github.com/mingfeima/op_bench-py), to reproduce, `./run.sh uniform.py`. Test machine is Xeon Skylake 6148 with 2*20 cores @2.40GHz

benchmark uses the metric of **numbers processed per second**, aka the throughput, the higher the better.

#### single socket

input size | original | this pr | speedup
-- | -- | -- | --
[1000] | 0.157 | 0.381 | 2.43
[10000] | 0.188 | 2.093 | 11.13
[100000] | 0.191 | 11.545 | 60.45
[1000000] | 0.191 | 27.207 | 142.45

#### single core

input size | original | this pr | speedup
-- | -- | -- | --
[1000] | 0.160 | 0.508 | 3.18
[10000] | 0.188 | 1.802 | 9.59
[100000] | 0.191 | 2.436 | 12.75
[1000000] | 0.191 | 2.517 | 13.18

### Validation
To prove the MKL VSL random generator as functional as ATen implementation, test cases are also provided in `uniform.py`.

The test case uses [T-test](http://geog.uoregon.edu/GeogR/topics/ttest.pdf) to verify the **mean** and **variance** of MKL VSL output aligns with theoretical **mean** and **variance** for [uniform distribution](https://en.wikipedia.org/wiki/Uniform_distribution_(continuous)):
```
for X~U(a, b)
E(X) = (a+b)/2
VAR(X) = E(X^2) - E(X)^2 = (b-a)^2 / 12
```

I do cross check the test cases with ATen implementation and MKL VSL implementation, both passed.

**Comment by XiaobingSuper** [comment]
#24781 Migrate uniform_ from the TH to Aten (CPU)

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit dc9d6546:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=dc9d654637eb0db559a1c709ee4fcd40861d511a).

### :detective: 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4050624) pytorch_windows_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=78561931) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=78561931))
<details>
<summary>
<code>CMake Error at torch_cuda_generated_THCTensorMaskedDouble.cu.obj.Release.cmake:281 (message):</code>
</summary>

```
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/advance.inl(33): error C2923: 'std::_Select<__formal>::_Apply': '_1' is not a valid template type argument for parameter '<unnamed-symbol>'
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/advance.inl(33): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/advance.inl(33): error C2144: syntax error: 'unknown-type' should be preceded by ')'
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/advance.inl(33): error C2144: syntax error: 'unknown-type' should be preceded by ';'
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/advance.inl(33): error C2238: unexpected token(s) preceding ';'
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/advance.inl(33): error C2059: syntax error: ')'
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/advance.inl(33): error C2988: unrecognizable template declaration/definition
C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include\thrust/detail/advance.inl(33): error C2059: syntax error: '<end Parse>'
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMaskedDouble.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMaskedDouble.cu.obj
CMake Error at torch_cuda_generated_THCTensorMaskedDouble.cu.obj.Release.cmake:281 (message):
  Error generating file
  C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMaskedDouble.cu.obj


:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathReduceBool.cu.obj -D generated_cubin_file:STRING=C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathReduceBool.cu.obj.cubin.txt -P C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorMathReduceBool.cu.obj.Release.cmake" 
-- Removing C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathReduceBool.cu.obj
"C:/Program Files/CMake/bin/cmake.exe" -E remove C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/./torch_cuda_generated_THCTensorMathReduceBool.cu.obj
-- Generating dependency file: C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/THC/generated/torch_cuda_generated_THCTensorMathReduceBool.cu.obj.NVCC-depend
C:/Users/circleci/project/aten/src/THCUNN -IC:/Users/circleci/project/aten/src/ATen/cuda -IC:/Users/circleci/project/build/caffe2/aten/src -IC:/Users/circleci/project/aten/../third_party/catch/single_include -IC:/Users/circleci/project/aten/src/ATen/.. -IC:/Users/circleci/project/build/caffe2/aten/src/ATen -IC:/Users/circleci/project/c10/cuda/../.. -IC:/Users/circleci/project/c10/../ -IC:/Users/circleci/project/caffe2/../torch/csrc/api -IC:/Users/circleci/project/caffe2/../torch/csrc/api/include
THCTensorMathReduceBool.cu 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: aten/src/ATen/native/Distributions.cpp [hunk 1]
```diff
@@ -179,6 +179,27 @@ Tensor& bernoulli_scalar_cpu_(Tensor& self, double p, Generator* gen) {
...
+  if (cpuinfo_initialize() && cpuinfo_vendor_intel == cpuinfo_get_processor(0)->core->vendor) {
+    uniform_mkl_stub(kCPU, self, from, to, gen);
+    return self;
+  }
+#endif
+  AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "uniform_cpu_", [&] {
+    CPUGenerator* generator = at::get_generator_or_default<CPUGenerator>(gen, detail::getDefaultCPUGenerator());
+    // See Note [Acquire lock when using random generators]
+    std::lock_guard<std::mutex> lock(generator->mutex_);
+    CPU_tensor_apply1<scalar_t>(
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/Distributions.cpp]: Please use `TensorIterator` and `cpu_serial_kernel ` see https://github.com/pytorch/pytorch/pull/32534

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 2]
```diff
@@ -311,6 +311,68 @@ void bernoulli_mkl_kernel(Tensor &self, const double p, Generator* gen) {
...
+    const int n, float* r, const float a, const float b) {
+  vsRngUniform( method, stream, n, r, a, b );
+}
+
+static inline void vslRngUniform(const int method, VSLStreamStatePtr stream,
+    const int n, double* r, const double a, const double b) {
+  vdRngUniform( method, stream, n, r, a, b );
+}
+
+void uniform_mkl_kernel(Tensor & self, const double from, const double to, Generator* gen) {
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp]: Do you need two copies of this code compiled (AVX on/off)?

**Review by VitalyFedyunin — CHANGES_REQUESTED** [review]
Please use TensorIterator, please rebase.

File: aten/src/ATen/native/Distributions.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/Distributions.cpp]: Done, `CPU_tensor_apply1` has been replaced by `TensorIterator`.

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 2 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp]: Not really, I just aligned `uniform_mkl_kernel` with `bernoulli_mkl_kernel` in the same manner.

**Review by VitalyFedyunin — DISMISSED** [review]

**Comment by VitalyFedyunin** [comment]
cc @pbelevich as it is related to random generators

**Comment by VitalyFedyunin** [comment]
I had no time to check in details, but errors looks like related to the introduced change and I cannot merge it as is.

**Comment by mingfeima** [comment]
> I had no time to check in details, but errors looks like related to the introduced change and I cannot merge it as is.

yes, i was aware of the errors are real... Somehow I have some trouble in understanding the test cases. Still WIP.

**Comment by mingfeima** [comment]
Remove `vsl` and use plain CPU implementation for now. Will put `vsl` part in following work.

**Review by VitalyFedyunin — APPROVED** [review]
Please rebase

**Comment by mingfeima** [comment]
Rebased

**Comment by VitalyFedyunin** [comment]
Sorry slipped through the cracks, please rebase.

**Comment by mingfeima** [comment]
rebased

**Comment by VitalyFedyunin** [comment]
I think we can close this one, as it is already ported.

**Pull request closed without merging**